### PR TITLE
A cleaner solution to grad sharing problem

### PR DESCRIPTION
**Pull request by apaszke**

### CODE REVIEW

**Comment by ikostrikov** [comment]
Thanks a lot!

**Comment by utanashati** [comment]
@apaszke Could you please explain the logic behind `ensure_shared_grads()`? I am reading it like "update the shared grads while they're None, stop when found a parameter that is not None". Is it meant only for the first iteration, when the `shared_model`’s grads are still None?

I did some logging of the parameters updates, and it’s a bit confusing.

First, after the initial round of updates, the `shared_model` gets updated right after the `.backward()`, which seems weird to me because the loss tensors all come from the local `model`. The only thing I can imagine is some kind of referencing to the `shared_model` when loading the `state_dict`.

Second, the `.step()` sometimes misses out an update from one of the workers. The sequence might be _pr1-backward — pr2-backward — pr1-step — pr2-step_, which only does `.step()` on the gradient accumulated by process 2. Shouldn’t we use lock to prevent this from happening?

For reference, take a look at the log and the code in the attachment.

[log.txt](https://github.com/ikostrikov/pytorch-a3c/files/3788212/log.txt)
[train.py.txt](https://github.com/ikostrikov/pytorch-a3c/files/3788213/train.py.txt)

**Comment by FengKaiyu** [comment]
@utanashati 
You may find the following link useful, which perfectly answers your two questions.
https://discuss.pytorch.org/t/grad-sharing-problem-in-a3c/10635
